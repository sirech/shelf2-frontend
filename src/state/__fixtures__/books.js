const books = () => (
  {
    entities: {
      books: {
        '4': {
          id: 4,
          title: 'Fear and Trembling',
          year: 2016,
          stars: 5,
          category: 'history'
        },
        '6': {
          id: 6,
          title: 'The Painted Veil',
          year: 2016,
          stars: 1,
          category: 'sociology'
        },
        '7': {
          id: 7,
          title: 'Things Fall Apart',
          year: 2016,
          stars: 1,
          category: 'other'
        },
        '8': {
          id: 8,
          title: 'Alone on a Wide, Wide Sea',
          year: 2016,
          stars: 2,
          category: 'sociology'
        },
        '9': {
          id: 9,
          title: 'I Know Why the Caged Bird Sings',
          year: 2016,
          stars: 2,
          category: 'other'
        },
        '10': {
          id: 10,
          title: 'Carrion Comfort',
          year: 2016,
          stars: 1,
          category: 'other'
        },
        '11': {
          id: 11,
          title: 'Time To Murder And Create',
          year: 2016,
          stars: 2,
          category: 'history'
        }
      }
    },
    result: [
      8,
      10,
      4,
      9,
      6,
      7,
      11
    ],
    activeYear: 2016
  }
)

export default books
