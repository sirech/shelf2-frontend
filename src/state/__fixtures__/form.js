const form = () => ({
  book: {
    title: '',
    year: 2017,
    description: '',
    stars: 1,
    category: 'software'
  },
  forms: {
    $form: {
      initialValue: {
        book: {
          title: '',
          year: 2017,
          description: '',
          stars: 1,
          category: 'software'
        }
      },
      focus: false,
      pending: false,
      pristine: true,
      submitted: false,
      submitFailed: false,
      retouched: false,
      touched: false,
      valid: true,
      validating: false,
      validated: false,
      validity: {},
      errors: {},
      intents: [],
      model: 'form',
      value: {
        book: {
          title: '',
          year: 2017,
          description: '',
          stars: 1,
          category: 'software'
        }
      }
    },
    book: {
      $form: {
        initialValue: {
          title: '',
          year: 2017,
          description: '',
          stars: 1,
          category: 'software'
        },
        focus: false,
        pending: false,
        pristine: true,
        submitted: false,
        submitFailed: false,
        retouched: false,
        touched: false,
        valid: true,
        validating: false,
        validated: false,
        validity: {},
        errors: {},
        intents: [],
        model: 'form.book',
        value: {
          title: '',
          year: 2017,
          description: '',
          stars: 1,
          category: 'software'
        }
      },
      title: {
        initialValue: '',
        focus: false,
        pending: false,
        pristine: false,
        submitted: false,
        submitFailed: false,
        retouched: false,
        touched: false,
        valid: true,
        validating: false,
        validated: false,
        validity: {},
        errors: {},
        intents: [],
        model: 'form.book.title',
        value: 'stuff'
      },
      year: {
        initialValue: 2017,
        focus: false,
        pending: false,
        pristine: true,
        submitted: false,
        submitFailed: false,
        retouched: false,
        touched: false,
        valid: true,
        validating: false,
        validated: false,
        validity: {},
        errors: {},
        intents: [],
        model: 'form.book.year',
        value: 2017
      },
      description: {
        initialValue: '',
        focus: false,
        pending: false,
        pristine: true,
        submitted: false,
        submitFailed: false,
        retouched: false,
        touched: false,
        valid: true,
        validating: false,
        validated: false,
        validity: {},
        errors: {},
        intents: [],
        model: 'form.book.description',
        value: ''
      },
      stars: {
        initialValue: 1,
        focus: false,
        pending: false,
        pristine: true,
        submitted: false,
        submitFailed: false,
        retouched: false,
        touched: false,
        valid: true,
        validating: false,
        validated: false,
        validity: {},
        errors: {},
        intents: [],
        model: 'form.book.stars',
        value: 1
      },
      category: {
        initialValue: 'software',
        focus: false,
        pending: false,
        pristine: true,
        submitted: false,
        submitFailed: false,
        retouched: false,
        touched: false,
        valid: true,
        validating: false,
        validated: false,
        validity: {},
        errors: {},
        intents: [],
        model: 'form.book.category',
        value: 'software'
      }
    }
  }
})

export default form
