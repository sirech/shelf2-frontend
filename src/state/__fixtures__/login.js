const login = () => ({
  authenticated: true,
  failed: false,
})

export default login
