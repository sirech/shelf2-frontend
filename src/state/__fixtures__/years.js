const years = () => (
  {
    entities: {
      years: {
        '2010': {
          year: 2010,
          count: 1
        },
        '2012': {
          year: 2012,
          count: 1
        },
        '2013': {
          year: 2013,
          count: 1
        },
        '2014': {
          year: 2014,
          count: 1
        },
        '2016': {
          year: 2016,
          count: 7
        }
      }
    },
    result: [
      2010,
      2012,
      2013,
      2014,
      2016
    ]
  }
)

export default years
