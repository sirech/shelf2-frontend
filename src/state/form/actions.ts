// @flow

import { fetch } from 'rest'

import { BOOK_CREATE_SUCCESS, BOOK_CREATE_FAIL } from './constants'

import type { BookForm, Book } from 'types'

const bookCreated = (book: Book) => ({
  type: BOOK_CREATE_SUCCESS,
  payload: book,
})

const bookCreationFailed = (error) => ({
  type: BOOK_CREATE_FAIL,
  payload: error,
})

export const create = (
  book: BookForm,
  history: Object,
  errorCallback: (string) => void
) => {
  return (dispatch: Dispatch) => {
    const url = '/books'
    const method = 'POST'
    const data = { book }

    return fetch(url, { method, data })
      .then((response) => dispatch(bookCreated(response.data)))
      .then(() => history.push('/books'))
      .catch((error) => {
        dispatch(bookCreationFailed(error.response.data))
        errorCallback(error.response.data)
      })
  }
}
