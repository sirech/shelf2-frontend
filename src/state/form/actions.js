// @flow

import { fetch } from 'rest'

import { BOOK_CREATE_SUCCESS, BOOK_CREATE_FAIL } from './constants'

import { actions as loginActions } from '../login'

import type { BookForm, Book } from 'types'

const bookCreated = (book: Book) => ({
  type: BOOK_CREATE_SUCCESS,
  payload: book,
})

const bookCreationFailed = error => ({
  type: BOOK_CREATE_FAIL,
  payload: error,
})

export const create = (book: BookForm, history: Object) => {
  return (dispatch: Dispatch) => {
    const url = '/books'
    const method = 'POST'
    const data = { book }

    return fetch(url, { method, data })
      .then(response => dispatch(bookCreated(response.data)))
      .then(() => history.push('/books'))
      .catch(error => {
        localStorage.removeItem('authToken')
        dispatch(bookCreationFailed(error.response.data))
        dispatch(loginActions.logoutSuccess())
      })
  }
}
