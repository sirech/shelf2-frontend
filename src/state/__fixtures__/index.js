import R from 'ramda'

import books from './books'
import years from './years'
import form from './form'
import login from './login'

const modal = () => ({
  opened: false
})

const search = R.pipe(books, R.omit(['activeYear']))

const state = () => {
  return {
    books: books(),
    search: search(),
    years: years(),
    form: form(),
    modal: modal(),
    login: login()
  }
}

export { books, search, years, form, login }
export default state
