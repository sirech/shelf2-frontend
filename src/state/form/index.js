import * as constants from './constants'
import * as actions from './actions'

export { constants, actions }
export { default as formSelector, bookSelector } from './selectors'
export { default as reducer } from './reducers'

export { modelName } from './utils'
