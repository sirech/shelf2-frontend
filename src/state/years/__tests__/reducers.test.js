import reducer from '../reducers'

import { constants } from '../../form'
import { factories } from 'test'
import { years } from '../../__fixtures__'

describe('years reducer', () => {
  it('returns the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  describe('handles BOOK_CREATE_SUCCESS', () => {
    it('it adds a new year entry', () => {
      const book = factories.book.build({ year: 2100 })
      expect(reducer(years(), { type: constants.BOOK_CREATE_SUCCESS, payload: book })).toMatchSnapshot()
    })

    it('increments the counter for existing years', () => {
      const book = factories.book.build({ year: 2010 })
      expect(reducer(years(), { type: constants.BOOK_CREATE_SUCCESS, payload: book })).toMatchSnapshot()
    })
  })
})
