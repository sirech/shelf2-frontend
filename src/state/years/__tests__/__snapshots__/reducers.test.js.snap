// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`years reducer handles BOOK_CREATE_SUCCESS increments the counter for existing years 1`] = `
Object {
  "entities": Object {
    "years": Object {
      "2010": Object {
        "count": 2,
        "year": 2010,
      },
      "2012": Object {
        "count": 1,
        "year": 2012,
      },
      "2013": Object {
        "count": 1,
        "year": 2013,
      },
      "2014": Object {
        "count": 1,
        "year": 2014,
      },
      "2016": Object {
        "count": 7,
        "year": 2016,
      },
    },
  },
  "result": Array [
    2010,
    2012,
    2013,
    2014,
    2016,
  ],
}
`;

exports[`years reducer handles BOOK_CREATE_SUCCESS it adds a new year entry 1`] = `
Object {
  "entities": Object {
    "years": Object {
      "2010": Object {
        "count": 1,
        "year": 2010,
      },
      "2012": Object {
        "count": 1,
        "year": 2012,
      },
      "2013": Object {
        "count": 1,
        "year": 2013,
      },
      "2014": Object {
        "count": 1,
        "year": 2014,
      },
      "2015": Object {
        "count": 1,
        "year": 2015,
      },
      "2016": Object {
        "count": 7,
        "year": 2016,
      },
    },
  },
  "result": Array [
    2010,
    2012,
    2013,
    2014,
    2015,
    2016,
  ],
}
`;

exports[`years reducer returns the initial state 1`] = `
Object {
  "entities": Object {
    "years": Object {},
  },
  "result": Array [],
}
`;
